File: metadata/src/application/dtos/edge_data_dto.rs
use super::PropertiesDataDto;
use crate::domain::models::{CreateEdgeData, EdgeData, EdgeDataId, NodeDataId};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

#[derive(Debug, Clone, Deserialize, ToSchema)]
pub struct CreateEdgeDataDto {
    pub key: String,
    pub from_node_data_id: NodeDataId,
    pub to_node_data_id: NodeDataId,
    pub properties: PropertiesDataDto,
}

impl CreateEdgeDataDto {
    pub fn into_domain(self) -> CreateEdgeData {
        CreateEdgeData {
            key: self.key,
            from_node_data_id: self.from_node_data_id,
            to_node_data_id: self.to_node_data_id,
            properties: self.properties.into(),
        }
    }
}

#[derive(Debug, Serialize, ToSchema)]
pub struct EdgeDataDto {
    pub edge_data_id: EdgeDataId,
    pub key: String,
    pub from_node_data_id: NodeDataId,
    pub to_node_data_id: NodeDataId,
    pub properties: PropertiesDataDto,
}

impl From<EdgeData> for EdgeDataDto {
    fn from(edge_data: EdgeData) -> Self {
        Self {
            edge_data_id: edge_data.edge_data_id,
            key: edge_data.key,
            from_node_data_id: edge_data.from_node_data_id,
            to_node_data_id: edge_data.to_node_data_id,
            properties: edge_data.properties.into(),
        }
    }
}

File: web-ui/lib/api/dtos/edge-data-dto.ts
import * as v from "valibot";
import { PropertiesDataDto } from "./property-data-dto";

export const CreateEdgeDataDto = v.object({
  key: v.string(),
  from_node_data_id: v.string(),
  to_node_data_id: v.string(),
  properties: PropertiesDataDto
});

export const EdgeDataDto = v.object({
  edge_data_id: v.string(),
  key: v.string(),
  from_node_data_id: v.string(),
  to_node_data_id: v.string(),
  properties: PropertiesDataDto
});
